### PR TITLE
Added `kubectl create clusterrole` command.

### PR DESCRIPTION
**Pull request by xingzhou**
Added `kubectl create clusterrole` command.

Fixed part of #39596 

**Special notes for your reviewer**:
@deads2k, please help to review this patch, thanks

**Release note**:
```
   Added one new command `kubectl create clusterrole` to help user create a single ClusterRole from command line.
```

### CODE REVIEW

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+func (c *CreateClusterRoleOptions) Complete(cmd *cobra.Command, args []string) error {
+	name, err := NameFromCommandArgs(cmd, args)
+	if err != nil {
+		return err
+	}
+	c.Name = name
+
+	// Remove duplicate verbs.
+	verbs := []string{}
+	for _, v := range c.Verbs {
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: Can you factor out some of this duplicate logic from the two commands?

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+		}
+
+		if arrayContains(validNonResourceVerbs, v) {
+			nonResourceVerbs = append(nonResourceVerbs, v)
+		}
+	}
+
+	clusterRole := &rbac.ClusterRole{}
+
+	// Create separate rule for each of the api group.
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: Creating the resource policy rules should be common code between create role and create cluster role.

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+		if err != nil {
+			return err
+		}
+		if !arrayContains(groupResourceMapping[resource.Group], resource.Resource) {
+			groupResourceMapping[resource.Group] = append(groupResourceMapping[resource.Group], resource.Resource)
+		}
+	}
+
+	// Split verbs to resource verbs and non-resource verbs
+	resourceVerbs, nonResourceVerbs := []string{}, []string{}
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: you aren't going to be able to distinguish

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+		},
+	}
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddDryRunFlag(cmd)
+	cmd.Flags().StringSliceVar(&c.Verbs, "verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSliceVar(&c.ResourceNames, "resource-name", []string{}, "resource in the white list that the rule applies to")
+	cmd.Flags().StringSliceVar(&c.NonResourceURLs, "non-resource-url", []string{}, "non-resource URL is a partial URL that a user should have access to")
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: Thinking about this, I suspect we actually want to defer this option until later.  We may even say its only on the `set` command.  The vast majority of usage is interested in resource access.

**Comment by deads2k** [comment]
I think narrowing the scope and DRYing it out a little bit will make this work out nicely.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-cli-pr-reviews

**Comment by xingzhou** [comment]
David, I understand that we want to keep `kubectl create` series command simple, but seems `nonresourceurl` is the only different options to role options that I can see. Do we plan to add it some day in the future or we just rely on `set` command to modify the existing cluster roles? Since if we remove `nonresourceurl`, in terms of options, there won't be much difference between `cluster role` and `role`, and most of the complete/validate/run code can be extract to a common place.

**Comment by deads2k** [comment]
> Do we plan to add it some day in the future or we just rely on set command to modify the existing cluster roles?

Yes, I think we will want to add it, but we need to consider how it interacts with things like verbs.  I think from a structure perspective, you could build the resource rules in a common fashion and then later add the non-resource rules, so we aren't boxed into a corner either way.

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+	if len(c.Resources) == 0 && len(c.NonResourceURLs) == 0 {
+		return fmt.Errorf("at least one resource or non-resource URL must be specified")
+	}
+
+	return nil
+}
+
+func (c *CreateClusterRoleOptions) RunCreateClusterRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
+	mapper, _ := f.Object()
```
Comment (AdoHe) [inline pkg/kubectl/cmd/create_clusterrole.go]: actually I would not expect to depends on cobra on the run logical.

Comment (xingzhou) [inline pkg/kubectl/cmd/create_clusterrole.go]: @AdoHe, any code patterns need to follow here? extract the arguments from cmd object and pass them in as parameters? wonder any examples in current commands

**Comment by xingzhou** [comment]
Have removed `non-resource-url` option now. Keep `CreateClusterRole` struct in case we insert `non-resource-url` in the future, PTAL. thx

File: pkg/kubectl/cmd/create_clusterrole.go (hunk 5 above)
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: > @AdoHe, any code patterns need to follow here? extract the arguments from cmd object and pass them in as parameters? wonder any examples in current commands

I missed this.  Yeah, `Complete` should gather this and set it in the options.

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddDryRunFlag(cmd)
+	cmd.Flags().StringSliceVar(&c.Verbs, "verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSliceVar(&c.ResourceNames, "resource-name", []string{}, "resource in the white list that the rule applies to")
+
+	return cmd
+}
+
+func (c *CreateClusterRoleOptions) RunCreateRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_clusterrole.go]: As @AdoHe pointed out.  Don't pass arguments here.  Instead, use `Complete` to gather the bits you need and store the on your options.

**Comment by xingzhou** [comment]
Have tried to move all the parameters from `Run` method to `CreateRoleOptions`, please take a look ,thanks.

File: pkg/kubectl/cmd/create_role.go [hunk 7]
```diff
@@ -50,20 +53,30 @@ type CreateRoleOptions struct {
 	Verbs         []string
 	Resources     []schema.GroupVersionResource
 	ResourceNames []string
+
+	DryRun       bool
+	OutputFormat string
```
Comment (deads2k) [inline pkg/kubectl/cmd/create_role.go]: At some point, I think we'll need to start wrapping this logic inside of the `PrintObject` function.  You don't need to mess with it here.

**Comment by deads2k** [comment]
@k8s-bot unit test this: issue #41885

**Comment by deads2k** [comment]
@k8s-bot gci gce e2e test this

**Comment by deads2k** [comment]
@k8s-bot non-cri e2e test this: issue #41889

File: pkg/kubectl/cmd/create_role.go (hunk 7 above)
Comment (xingzhou) [inline pkg/kubectl/cmd/create_role.go]: I agree, those are just common command options, not specific to this one. Do we have plan now, please let me know if I can help, thanks!

**Pull request merged**